diff -up gnome-panel-2.28.0/gnome-panel/panel-util.c.dir-prefix gnome-panel-2.28.0/gnome-panel/panel-util.c
--- gnome-panel-2.28.0/gnome-panel/panel-util.c.dir-prefix	2010-03-21 20:14:19.014530854 -0400
+++ gnome-panel-2.28.0/gnome-panel/panel-util.c	2010-03-21 20:17:11.275527640 -0400
@@ -447,8 +447,16 @@ panel_lock_screen (GdkScreen *screen)
 static char *
 panel_launcher_get_personal_path (void)
 {
-	return g_build_filename (g_get_home_dir (), ".gnome2",
-				 PANEL_LAUNCHER_PERSONAL_PATH, NULL);
+        const char *override;
+
+        override = g_getenv ("GNOME22_USER_DIR");
+
+        if (override)
+               return g_build_filename (override,
+                                        PANEL_LAUNCHER_PERSONAL_PATH, NULL);
+        else
+                return g_build_filename (g_get_home_dir (), ".gnome2",
+                                         PANEL_LAUNCHER_PERSONAL_PATH, NULL);
 }
 
 gboolean
